lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,660 @@
+    share_reserves : FixedPoint
+        The amount of share reserves in the Hyperdrive pool.
+    bonds_in : FixedPoint
+        The amount of bonds entering the pool.
+    time_stretch : FixedPoint
+        The time stretch factor.
+    curve_fee : FixedPoint
+        The curve fee, as a percentage of the price discount, from pool config.
+    gov_fee : FixedPoint
+        The governance fee, as a percentage of the flat+curve fee, from pool config.
+    one_block_return : FixedPoint, optional
+        An estimate of the variable return expected for the next block, by default None.
+    """
+    # pylint: disable=too-many-arguments
+    if one_block_return is None:
+        one_block_return = FixedPoint(1)
+    k_t = calc_k_local(
+        share_price,
+        initial_share_price,
+        share_reserves,
+        bond_reserves,
+        time_stretch,
+    )
+    y_term = (bond_reserves + bonds_in) ** (FixedPoint(1) - time_stretch)
+    z_val = (k_t - y_term) / (share_price / initial_share_price)
+    z_val = z_val ** (FixedPoint(1) / (FixedPoint(1) - time_stretch))
+    z_val /= initial_share_price
+    # z_val *= one_block_return  # unclear if this helps improve accuracy
+    spot_price = calc_spot_price_local(initial_share_price, share_reserves, FixedPoint(0), bond_reserves, time_stretch)
+    price_discount = FixedPoint(1) - spot_price
+    amount_in_shares = max(FixedPoint(0), share_reserves - z_val)
+    curve_fee_rate = price_discount * curve_fee
+    curve_fee_amount_in_shares = amount_in_shares * curve_fee_rate
+    gov_fee_amount_in_shares = curve_fee_amount_in_shares * gov_fee
+    # applying fee means you get LESS shares out for the same amount of bonds IN
+    amount_to_user_in_shares = amount_in_shares - curve_fee_amount_in_shares
+    return amount_to_user_in_shares, curve_fee_amount_in_shares, gov_fee_amount_in_shares
+
+
+# TODO: switch over to using function in the SDK
+def calc_reserves_to_hit_target_rate(
+    target_rate: FixedPoint, interface: HyperdriveInterface
+) -> tuple[FixedPoint, FixedPoint]:
+    """Calculate the bonds and shares needed to hit the target fixed rate.
+
+    Arguments
+    ---------
+    target_rate : FixedPoint
+        The target rate the pool will have after the calculated change in bonds and shares.
+    interface : HyperdriveInterface
+        The Hyperdrive API interface object.
+
+    Returns
+    -------
+    tuple[FixedPoint, FixedPoint] containing:
+        total_shares_needed : FixedPoint
+            Total amount of shares needed to be added into the pool to hit the target rate.
+        total_bonds_needed : FixedPoint
+            Total amount of bonds needed to be added into the pool to hit the target rate.
+    """
+    # variables
+    predicted_rate = FixedPoint(0)
+    pool_config = interface.pool_config.copy()
+    pool_info = interface.pool_info.copy()
+
+    variable_rate = interface.variable_rate
+    one_block_return = (FixedPoint(1) + variable_rate) ** (
```

Comment:
pull this out since it's not used, we can add it in later when we discuss how to make it smarter